contracts/contracts/ERC677/starkgate_token.cairo

```diff
@@ -0,0 +1,195 @@
+%lang starknet
+
+from starkware.cairo.common.cairo_builtins import HashBuiltin
+from starkware.cairo.common.alloc import alloc
+from starkware.cairo.common.math import assert_not_zero, assert_le
+from starkware.cairo.common.uint256 import (
+    Uint256,
+    uint256_add,
+    uint256_check,
+    uint256_le,
+    uint256_lt,
+    uint256_sub,
+)
+from starkware.starknet.common.syscalls import get_caller_address
+from starkware.cairo.common.bool import FALSE, TRUE
+from openzeppelin.introspection.erc165.library import ERC165
+from contracts.starkware.starknet.std_contracts.ERC20.ERC20_base import (
+    ERC20_allowances,
+    ERC20_approve,
+    ERC20_burn,
+    ERC20_initializer,
+    ERC20_mint,
+    ERC20_transfer,
+    allowance,
+    balanceOf,
+    decimals,
+    name,
+    symbol,
+    totalSupply,
+)
+from contracts.starkware.starknet.std_contracts.ERC20.permitted import (
+    permitted_initializer,
+    permitted_minter,
+    permitted_minter_only,
+    permittedMinter,
+)
+from contracts.starkware.starknet.std_contracts.upgradability_proxy.initializable import (
+    initialized,
+    set_initialized,
+)
+from contracts.ERC677.library import ERC677
+from contracts.ERC677.interfaces.IERC677Receiver import IERC677Receiver
+
+const CONTRACT_IDENTITY = 'ERC20'
+const CONTRACT_VERSION = 1
+
+@view
+func get_version() -> (version : felt):
+    return (version=CONTRACT_VERSION)
+end
+
+@view
+func get_identity() -> (identity : felt):
+    return (identity=CONTRACT_IDENTITY)
+end
+
+# Constructor (as initializer).
+
+@external
+func initialize{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    init_vector_len : felt, init_vector : felt*
+):
+    set_initialized()
+    # We expect the init vector to be [name , symbol , decimals , minter_address].
+    with_attr error_message("ILLEGAL_INIT_SIZE"):
+        assert init_vector_len = 4
+    end
+
+    let name = [init_vector]
+    let symbol = [init_vector + 1]
+    let decimals = [init_vector + 2]
+    ERC20_initializer(name, symbol, decimals)
+    let minter_address = [init_vector + 3]
+    permitted_initializer(minter_address)
+    return ()
+end
+
+# Externals.
+
+@external
+func transfer{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    recipient : felt, amount : Uint256
+) -> (success : felt):
+    let (sender) = get_caller_address()
+    ERC20_transfer(sender, recipient, amount)
+
+    # Cairo equivalent to 'return (true)'
+    return (1)
+end
+
+@external
+func transferFrom{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    sender : felt, recipient : felt, amount : Uint256
+) -> (success : felt):
+    alloc_locals
```

Comment:
Wow! Why do we duplicate all this code?

Please take a look at OZ ERC20 presets, for example: https://github.com/OpenZeppelin/cairo-contracts/blob/main/src/openzeppelin/token/erc20/presets/ERC20.cairo

It makes no sense to duplicate, review and audit all this code that already exists (!?). We want to build on top, and extend with minimal amount of code. 

If the source, Starkgate ERC20 contract, is not prepared to be easily extended while the OZ ERC20 is, then let's not use it. Let's use the OZ implementation and then extend that one with the Starkgate expected interface: https://github.com/starkware-libs/starkgate-contracts/blob/main/src/starkware/starknet/std_contracts/ERC20/mintable_token_interface.cairo

Please restructure code like this:

- `/ERC677/library.cairo` - base functionality
- `/ERC677/presets/ERC677.cairo` - OZ ERC20 + CL ERC677
- `/ERC677/presets/ERC677Mintable.cairo` - OZ ERC20Mintable + CL ERC677
- `/starkgate/presets/ERC677.cairo` - OZ ERC20Mintable + CL ERC677 + StarkGate IMintableToken impl

Contract naming is important, path naming (context) is important, how do we structure our code is important! Let's try to restructure this.

Reply:
@qperrot Let's use `ERC20Upgradeable` preset from OpenZeppelin https://github.com/OpenZeppelin/cairo-contracts/blob/main/src/openzeppelin/token/erc20/presets/ERC20Upgradeable.cairo then

```cairo
from starkware.starknet.std_contracts.ERC20.permitted import (
    permitted_initializer,
    permitted_minter,
    permitted_minter_only,
    permittedMinter,
)
```

And use them: https://github.com/starkware-libs/starkgate-contracts/blob/c08863a1f08226c09f1d0748124192e848d73db9/src/starkware/starknet/std_contracts/ERC20/ERC20.cairo#L160-L184

That way we are using the OZ token, but we extend it with the interface required by starkgate.

Reply:
@krebernisak we discussed with @archseer  and decided to skip the `presets` altogether and make a LinkToken: `OZ ERC20Upgradable + ERC677 + Starkgate IMintableToken (that permissible.cairo)`